File: csharp/LogicAndTrick.WikiCodeParser/Elements/QuoteElement.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Elements
{
    public class QuoteElement : Element
    {
        private static readonly Regex OpenQuote = new Regex(@"^\[quote(?:(?: name)?=([^]]*))?\]", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private const int CloseQuoteLength = 8;

        public override bool Matches(Lines lines)
        {
            var value = lines.Value().Trim();
            return value.Length > 6 && value.StartsWith("[quote") && OpenQuote.IsMatch(value);
        }

        public override INode Consume(Parser parser, ParseData data, Lines lines, string scope)
        {
            var current = lines.Current();
            var arr = new List<string>();

            var line = lines.Value().Trim();
            var res = OpenQuote.Match(line);
            if (!res.Success)
            {
                lines.SetCurrent(current);
                return null;
            }

            var text = BalanceQuotes(lines, out var author, out var postfix);
            if (text == null)
            {
                lines.SetCurrent(current);
                return null;
            }

            var before = "<blockquote>";
            var plainBefore = "[quote]\n";
            if (!string.IsNullOrWhiteSpace(author))
            {
                before += "<strong class=\"quote-name\">" + author + " said:</strong><br/>";
                plainBefore = author + " said: " + plainBefore;
            }

            var node = new HtmlNode(before, parser.ParseElements(data, text, scope), "</blockquote>")
            {
                PlainBefore = plainBefore,
                PlainAfter = "\n[/quote]",
                IsBlockNode = true,
            };
            if (!string.IsNullOrWhiteSpace(postfix))
            {
                return new NodeCollection(node, parser.ParseTags(data, postfix, scope, TagParseContext.Inline));
            }
            return node;
        }

        public static string BalanceQuotes(Lines lines, out string name, out string postfix)
        {
            name = null;
            postfix = null;

            var line = lines.Value().TrimStart();
            var openMat = OpenQuote.Match(line);
            if (!openMat.Success) return null;
            if (openMat.Groups[1].Success) name = openMat.Groups[1].Value;

            line = line.Substring(openMat.Length);
            var arr = new List<string>();
            var currentLevel = 1;
            do
            {
                var idx = 0;
                do
                {
                    openMat = OpenQuote.Match(line, idx);
                    var openMatIdx = openMat.Success ? openMat.Index : -1;
                    var closeMatIdx = line.IndexOf("[/quote]", idx, StringComparison.InvariantCultureIgnoreCase);

                    if (openMatIdx >= 0 && (closeMatIdx < 0 || closeMatIdx > openMatIdx))
                    {
                        // Open quote
                        currentLevel++;
                        idx = openMat.Index + openMat.Length;
                    }
                    else if (closeMatIdx >= 0)
                    {
                        // Close quote
                        currentLevel--;
                        if (currentLevel == 0)
                        {
                            if (line.Length > closeMatIdx + CloseQuoteLength) postfix = line.Substring(closeMatIdx + CloseQuoteLength);
                            arr.Add(line.Substring(0, closeMatIdx));
                            return string.Join("\n", arr);
                        }
                        idx = closeMatIdx + CloseQuoteLength;
                    }
                    else
                    {
                        arr.Add(line);
                        break;
                    }
                } while (true);

                if (lines.Next()) line = lines.Value();
                else break;
            } while (true);

            return null;
        }
    }
}

File: javascript/src/Elements/QuoteElement.ts
import { Parser } from '..';
import { Lines } from '../Lines';
import { HtmlNode } from '../Nodes/HtmlNode';
import { INode } from '../Nodes/INode';
import { NodeCollection } from '../Nodes/NodeCollection';
import { ParseData } from '../ParseData';
import { TagParseContext } from '../TagParseContext';
import { Element } from './Element';

export class QuoteElement extends Element {
    private static OpenQuote = /^\[quote(?:(?: name)?=([^\]]*))?\]/i;
    private static CloseQuoteLength = 8;

    Matches(lines: Lines): boolean {
        const value = lines.Value().trim();
        return value.length > 6 && value.toLowerCase().startsWith("[quote") && QuoteElement.OpenQuote.test(value);
    }

    public Consume(parser: Parser, data: ParseData, lines: Lines, scope: string): INode {
        const current = lines.Current();

        let line = lines.Value().trim();
        const res = QuoteElement.OpenQuote.exec(line);
        if (!res) {
            lines.SetCurrent(current);
            return null;
        }

        const { text, author, postfix } = QuoteElement.BalanceQuotes(lines);
        if (!text) {
            lines.SetCurrent(current);
            return null;
        }

        let before = "<blockquote>";
        let plainBefore = "[quote]\n";
        if (author) {
            before += `<strong class="quote-name">${author} said:</strong><br/>`;
            plainBefore = `${author} said: ${plainBefore}`;
        }

        const node = new HtmlNode(before, parser.ParseElements(data, text, scope), "</blockquote>");
        node.PlainBefore = plainBefore;
        node.PlainAfter = "\n[/quote]";
        node.IsBlockNode = true;

        if (postfix) {
            return new NodeCollection(node, parser.ParseTags(data, postfix, scope, TagParseContext.Inline));
        }
        return node;
    }

    public static BalanceQuotes(lines: Lines): { text: string | null, author: string | null, postfix: string | null } {
        let name = null;
        let postfix = null;
        
        const openQuote = new RegExp(QuoteElement.OpenQuote, 'iy');

        let line = lines.Value().trimStart();
        let openMat = openQuote.exec(line);
        if (!openMat) return { text: null, author: name, postfix };

        if (openMat[1]) name = openMat[1];

        line = line.substring(openMat[0].length);
        const arr = [];
        let currentLevel = 1;

        do {
            let idx = 0;
            do {
                openQuote.lastIndex = idx;
                openMat = openQuote.exec(line);
                const openMatIdx = openMat ? openMat.index : -1;
                const closeMatIdx = line.toLowerCase().indexOf("[/quote]", idx);

                if (openMatIdx >= 0 && (closeMatIdx < 0 || closeMatIdx > openMatIdx)) {
                    // Open quote
                    currentLevel++;
                    idx = openMat.index + openMat[0].length;
                } else if (closeMatIdx >= 0) {
                    // Close quote
                    currentLevel--;
                    if (currentLevel === 0) {
                        if (line.length > closeMatIdx + QuoteElement.CloseQuoteLength) postfix = line.substring(closeMatIdx + QuoteElement.CloseQuoteLength);
                        arr.push(line.substring(0, closeMatIdx));
                        return { text: arr.join("\n"), author: name, postfix };
                    }
                    idx = closeMatIdx + QuoteElement.CloseQuoteLength;
                } else {
                    arr.push(line);
                    break;
                }
            } while (true);

            if (lines.Next()) line = lines.Value();
            else break;
        } while (true);

        return { text: null, author: name, postfix };
    }
}
